### PR TITLE
r/kinesis_analytics_application: new resource

### PR DESCRIPTION
**Pull request by kl4w**
Still work in progress right now, still need to support `Outputs`: https://docs.aws.amazon.com/kinesisanalytics/latest/dev/API_CreateApplication.html#analytics-CreateApplication-request-Outputs

Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/536

Changes proposed in this pull request:

* create new kinesis_analytics_application resource

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSKinesisAnalyticsApplication_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSKinesisAnalyticsApplication_ -timeout 120m
=== RUN   TestAccAWSKinesisAnalyticsApplication_basic
--- PASS: TestAccAWSKinesisAnalyticsApplication_basic (13.58s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_update
--- PASS: TestAccAWSKinesisAnalyticsApplication_update (21.21s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions (67.57s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions (70.95s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream (131.00s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputUpdateKinesisStream (212.85s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	517.185s
```

### CODE REVIEW

File: aws/resource_aws_kinesis_analytics_application.go [hunk 1]
```diff
@@ -0,0 +1,868 @@
...
+		if len(oldLoggingOptions.([]interface{})) == 0 && len(newLoggingOptions.([]interface{})) > 0 {
+			if v, ok := d.GetOk("cloudwatch_logging_options"); ok {
+				clo := v.([]interface{})[0].(map[string]interface{})
+				cloudwatchLoggingOption := createCloudwatchLoggingOption(clo)
+				addOpts := &kinesisanalytics.AddApplicationCloudWatchLoggingOptionInput{
+					ApplicationName:             aws.String(name),
+					CurrentApplicationVersionId: aws.Int64(int64(version)),
+					CloudWatchLoggingOption:     cloudwatchLoggingOption,
+				}
+				conn.AddApplicationCloudWatchLoggingOption(addOpts)
```
Comment (kl4w) [inline aws/resource_aws_kinesis_analytics_application.go]: Not really sure the best way to handle this but adding new cloudwatch logging options cannot be part of the `UpdateApplication` call.

File: aws/resource_aws_kinesis_analytics_application_test.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+          }
+        }
+      }
+    }
+  }
+}
+`, streamName, rInt)
+}
+
+func fulfillSleep() resource.TestCheckFunc {
```
Comment (kl4w) [inline aws/resource_aws_kinesis_analytics_application_test.go]: I put in a sleep here because I found that the trust policy for the IAM role needs to propagate before referencing the role in the tests

File: vendor/vendor.json [hunk 3]
```diff
@@ -750,6 +750,14 @@
 			"version": "v1.14.33",
 			"versionExact": "v1.14.33"
 		},
+		{
+			"checksumSHA1": "nUdxsOW9jg+m+sWZYPu7oX9rZo8=",
+			"path": "github.com/aws/aws-sdk-go/service/kinesisanalytics",
```
Comment (bflad) [inline vendor/vendor.json]: Can you please separate the vendoring into its own PR? We do this because these dependencies are a fast moving target, it makes the PRs a little more manageable, and we can quickly merge it in. FYI `master` is already AWS SDK Go `v1.15.4`.

https://github.com/terraform-providers/terraform-provider-aws/blob/2c3eaaef70430fa462a5750825b2a339e910348a/vendor/vendor.json#L41-L48

Comment (kl4w) [reply vendor/vendor.json]: Absolutely. I've created the PR here: https://github.com/terraform-providers/terraform-provider-aws/pull/5461

File: aws/resource_aws_kinesis_analytics_application_test.go (hunk 2 above)
Comment (bflad) [reply aws/resource_aws_kinesis_analytics_application_test.go]: Hi @kl4w 👋  We prefer that the resource usually handles this type of logic by retrying on the specific error message(s) that are returned for a minute or two. e.g.

```go
// Handle IAM eventual consistency
err := resource.Retry(1*time.Minute, func() *resource.RetryError {
  // CreateApplication as an example
  err := conn.CreateApplication(input)
  if err != nil {
    // kinesisanalytics.ErrCodeInvalidArgumentException as an example
    if isAWSErr(err, kinesisanalytics.ErrCodeInvalidArgumentException, "some IAM-related error message, e.g. not authorized") {
      return resource.RetryableError(err)
    }
    return resource.NonRetryableError(err)
  }
  return nil
})
if err != nil {
  return fmt.Errorf("error creating application: %s", err)
}
```

**Comment by kl4w** [comment]
```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSKinesisAnalyticsApplication_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSKinesisAnalyticsApplication_ -timeout 120m
=== RUN   TestAccAWSKinesisAnalyticsApplication_basic
--- PASS: TestAccAWSKinesisAnalyticsApplication_basic (10.75s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_update
--- PASS: TestAccAWSKinesisAnalyticsApplication_update (14.77s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions (25.60s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions (37.05s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream (86.94s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsAdd
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsAdd (94.82s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsUpdateKinesisStream (170.57s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsKinesisStream (88.25s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsAdd
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsAdd (94.47s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsUpdateKinesisStream (182.43s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_referenceDataSource
--- PASS: TestAccAWSKinesisAnalyticsApplication_referenceDataSource (38.51s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_referenceDataSourceUpdate
--- PASS: TestAccAWSKinesisAnalyticsApplication_referenceDataSourceUpdate (59.21s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	903.415s
```

File: aws/resource_aws_kinesis_analytics_application.go [hunk 4]
```diff
@@ -0,0 +1,1521 @@
...
+func resourceAwsKinesisAnalyticsApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kinesisanalyticsconn
+	name := d.Get("name").(string)
+
+	describeOpts := &kinesisanalytics.DescribeApplicationInput{
+		ApplicationName: aws.String(name),
+	}
+	resp, err := conn.DescribeApplication(describeOpts)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: Nit: This error handling logic can be simplified with:

```go
if isAWSErr(err, kinesisanalytics.ErrCodeResourceNotFoundException, "") {
  log.Printf("[WARN] Kinesis Analytics Application (%d) not found, removing from state", d.Id())
  d.SetId("")
  return nil
}
if err != nil {
  return fmt.Errorf("error reading Kinesis Analytics Application (%s): %s", d.Id(), err)
}
```

File: aws/resource_aws_kinesis_analytics_application.go [hunk 5]
```diff
@@ -0,0 +1,1521 @@
...
+			if awsErr.Code() == "ResourceNotFoundException" {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("[WARN] Error reading Kinesis Analytics Application: \"%s\", code: \"%s\"", awsErr.Message(), awsErr.Code())
+		}
+		return err
+	}
+
+	d.SetId(aws.StringValue(resp.ApplicationDetail.ApplicationARN))
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: `d.SetId()` should be called during the `Create` function unless the `Update` function will change the ARN.

File: aws/resource_aws_kinesis_analytics_application.go [hunk 6]
```diff
@@ -0,0 +1,1521 @@
...
+		ApplicationName: aws.String(name),
+	}
+
+	if v, ok := d.GetOk("code"); ok && v.(string) != "" {
+		createOpts.ApplicationCode = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("cloudwatch_logging_options"); ok {
+		clo := v.([]interface{})[0].(map[string]interface{})
+		cloudwatchLoggingOption := createCloudwatchLoggingOption(clo)
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: Since this project uses a very flat Go package structure, we prefer to include the AWS service name in the function name. We also generally call these functions "expand" functions. e.g. `expandKinesisAnalyticsCloudwatchLoggingOption`

Same for the below `createX` function names.

File: aws/resource_aws_kinesis_analytics_application.go [hunk 7]
```diff
@@ -0,0 +1,1521 @@
...
+	d.Set("name", aws.StringValue(resp.ApplicationDetail.ApplicationName))
+	d.Set("arn", aws.StringValue(resp.ApplicationDetail.ApplicationARN))
+	d.Set("code", aws.StringValue(resp.ApplicationDetail.ApplicationCode))
+	d.Set("create_timestamp", aws.TimeValue(resp.ApplicationDetail.CreateTimestamp).Format(time.RFC3339))
+	d.Set("description", aws.StringValue(resp.ApplicationDetail.ApplicationDescription))
+	d.Set("last_update_timestamp", aws.TimeValue(resp.ApplicationDetail.LastUpdateTimestamp).Format(time.RFC3339))
+	d.Set("status", aws.StringValue(resp.ApplicationDetail.ApplicationStatus))
+	d.Set("version", int(aws.Int64Value(resp.ApplicationDetail.ApplicationVersionId)))
+
+	if err := d.Set("cloudwatch_logging_options", getCloudwatchLoggingOptions(resp.ApplicationDetail.CloudWatchLoggingOptionDescriptions)); err != nil {
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: Since this project uses a very flat Go package structure, we prefer to include the AWS service name in the function name. We also generally call these functions "flatten" functions. e.g. `flattenKinesisAnalyticsCloudwatchLoggingOptions`

Same for the below `getX` function names.

File: aws/resource_aws_kinesis_analytics_application.go [hunk 8]
```diff
@@ -0,0 +1,1521 @@
...
+		return err
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: Nit: Its generally helpful to provide the operator (and code maintainers) some context about the error message they are receiving, e.g. `return fmt.Errorf("error setting cloudwatch_logging_options: %s", err)`

Same for below `d.Set()` error handlers as well. 👍

File: aws/resource_aws_kinesis_analytics_application.go [hunk 9]
```diff
@@ -0,0 +1,1521 @@
...
+		return parseErr
+	}
+
+	log.Printf("[DEBUG] Kinesis Analytics Application destroy: %v", d.Id())
+	deleteOpts := &kinesisanalytics.DeleteApplicationInput{
+		ApplicationName: aws.String(name),
+		CreateTimestamp: aws.Time(createTimestamp),
+	}
+	_, deleteErr := conn.DeleteApplication(deleteOpts)
+	if deleteErr != nil {
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: If the application was previously deleted (e.g. outside Terraform), we should return without error:

```go
if isAWSErr(deleteErr, kinesisanalytics.ErrCodeResourceNotFoundException, "") {
  return nil
}
if deleteErr != nil {
  return fmt.Errorf("error deleting Kinesis Analytics Application (%s): %s", d.Id(), deleteErr)
}
```

File: aws/resource_aws_kinesis_analytics_application.go [hunk 10]
```diff
@@ -0,0 +1,1521 @@
...
+		return deleteErr
+	}
+
+	log.Printf("[DEBUG] Kinesis Analytics Application deleted: %v", d.Id())
+	return nil
+}
+
+func createCloudwatchLoggingOption(clo map[string]interface{}) *kinesisanalytics.CloudWatchLoggingOption {
+	cloudwatchLoggingOption := &kinesisanalytics.CloudWatchLoggingOption{
+		LogStreamARN: aws.String(clo["log_stream"].(string)),
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application.go]: We should prefer to stick to the API (and more clear) attribute naming for any of these `ARN` parameters that accept an ARN, e.g. `log_stream_arn`

This applies to a large amount of the attributes below as well so I'll withhold from commenting on each of them.

File: aws/resource_aws_kinesis_analytics_application_test.go [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+					resource.TestCheckResourceAttr(resName, "version", "1"),
+				),
+			},
+			{
+				Config: thirdStep,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckKinesisAnalyticsApplicationExists(resName, &application),
+					resource.TestCheckResourceAttr(resName, "version", "2"),
+					resource.TestCheckResourceAttr(resName, "cloudwatch_logging_options.#", "1"),
+					resource.TestMatchResourceAttr(resName, "cloudwatch_logging_options.0.log_stream", streamRe),
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application_test.go]: Instead of regexp matching the log stream ARN, you may find it easier to directly check the resource information via:

```go
resource.TestCheckResourceAttrPair(resName, "cloudwatch_logging_options.0.log_stream_arn", "aws_cloudwatch_log_stream.test", "arn"),
```

This applies to the other below checks as well 😄

File: aws/resource_aws_kinesis_analytics_application_test.go [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+		describeOpts := &kinesisanalytics.DescribeApplicationInput{
+			ApplicationName: aws.String(rs.Primary.Attributes["name"]),
+		}
+		resp, err := conn.DescribeApplication(describeOpts)
+		if err == nil {
+			if resp.ApplicationDetail != nil && *resp.ApplicationDetail.ApplicationStatus != kinesisanalytics.ApplicationStatusDeleting {
+				return fmt.Errorf("Error: Application still exists")
+			}
+		}
+		return nil
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application_test.go]: We should remove this `return nil` to ensure this function can check all configured `aws_kinesis_analytics_application` resources.

File: aws/resource_aws_kinesis_analytics_application_test.go [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+		if rs.Type != "aws_kinesis_analytics_application" {
+			continue
+		}
+		conn := testAccProvider.Meta().(*AWSClient).kinesisanalyticsconn
+		describeOpts := &kinesisanalytics.DescribeApplicationInput{
+			ApplicationName: aws.String(rs.Primary.Attributes["name"]),
+		}
+		resp, err := conn.DescribeApplication(describeOpts)
+		if err == nil {
+			if resp.ApplicationDetail != nil && *resp.ApplicationDetail.ApplicationStatus != kinesisanalytics.ApplicationStatusDeleting {
```
Comment (bflad) [inline aws/resource_aws_kinesis_analytics_application_test.go]: If the application does not delete immediately, should we be waiting for the deletion to complete in the `Delete` function? 🤔

**Comment by kl4w** [comment]
@bflad I've made changes as requested and rebased. test output:
```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSKinesisAnalyticsApplication_'
==> Fixing source code with gofmt...
gofmt -s -w ./aws
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSKinesisAnalyticsApplication_ -timeout 120m
=== RUN   TestAccAWSKinesisAnalyticsApplication_basic
--- PASS: TestAccAWSKinesisAnalyticsApplication_basic (12.64s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_update
--- PASS: TestAccAWSKinesisAnalyticsApplication_update (17.65s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions (27.30s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions (37.63s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream (88.75s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsAdd
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsAdd (97.11s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsUpdateKinesisStream (173.38s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsKinesisStream (90.16s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsAdd
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsAdd (96.73s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsUpdateKinesisStream (166.41s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_referenceDataSource
--- PASS: TestAccAWSKinesisAnalyticsApplication_referenceDataSource (29.32s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_referenceDataSourceUpdate
--- PASS: TestAccAWSKinesisAnalyticsApplication_referenceDataSourceUpdate (52.38s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	889.499s
```

**Comment by kl4w** [comment]
@bflad made changes as requested. test output:
```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSKinesisAnalyticsApplication_'
==> Fixing source code with gofmt...
gofmt -s -w ./aws
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSKinesisAnalyticsApplication_ -timeout 120m
=== RUN   TestAccAWSKinesisAnalyticsApplication_basic
--- PASS: TestAccAWSKinesisAnalyticsApplication_basic (11.24s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_update
--- PASS: TestAccAWSKinesisAnalyticsApplication_update (17.09s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_addCloudwatchLoggingOptions (26.86s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions
--- PASS: TestAccAWSKinesisAnalyticsApplication_updateCloudwatchLoggingOptions (37.94s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsKinesisStream (88.39s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsAdd
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsAdd (96.66s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_inputsUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_inputsUpdateKinesisStream (183.16s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsKinesisStream (89.69s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsAdd
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsAdd (96.81s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_outputsUpdateKinesisStream
--- PASS: TestAccAWSKinesisAnalyticsApplication_outputsUpdateKinesisStream (173.10s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_referenceDataSource
--- PASS: TestAccAWSKinesisAnalyticsApplication_referenceDataSource (28.78s)
=== RUN   TestAccAWSKinesisAnalyticsApplication_referenceDataSourceUpdate
--- PASS: TestAccAWSKinesisAnalyticsApplication_referenceDataSourceUpdate (49.67s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	899.442s
```

**Comment by bflad** [comment]
This has been released in [version 1.43.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1430-november-07-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by arturoherrero** [comment]
Hello @kl4w @bflad. Thank you very much for this contribution!

I created a `aws_kinesis_analytics_application` resource and everything is working as expected. The only problem I can't solve is how to start the Kinesis Analytics application.
https://docs.aws.amazon.com/kinesisanalytics/latest/dev/API_StartApplication.html

At the moment I can go directly to the UI and click the button to start the application:

screenshot 2019-01-29 at 09 51 28

I guess I can start the application using Terraform, maybe I have to configure that action or something else. Am I missing something?

**Comment by kl4w** [comment]
@arturoherrero you are correct, there's no functionality to start the application right now on creation. This probably could implemented by changing the `status` attribute: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_kinesis_analytics_application.go#L55-L58 and calling `StartApplication` post-creation though. Mind submitting a feature request if there's not one already?

**Comment by netrebel** [comment]
Hi @kl4w,

I'm trying to add an additional output for the `error_stream` created by Kinesis Analytics by default to handle runtime [errors](https://docs.aws.amazon.com/kinesisanalytics/latest/dev/error-handling.html).

The block in `aws_kinesis_analytics_application` looks like this:

```
outputs = [{
            name = "DESTINATION_SQL_STREAM"
            schema {
                record_format_type = "JSON"
            }
            lambda {
                resource_arn = "${aws_lambda_function.estimator.arn}"
                role_arn = "${aws_iam_role.lambda_execution.arn}"
            }
        },
        {
            name = "error_stream"
            schema {
                record_format_type = "JSON"
            }
            lambda {
                resource_arn = "${aws_lambda_function.splunk-logger.arn}"
                role_arn = "${aws_iam_role.lambda_execution.arn}"
            }
        }
    ]
```

But the corresponding block in the state file does not reflect that stream:

```
outputs.#: "1",
outputs.0.id: "1.1",
outputs.0.kinesis_firehose.#: "0",
outputs.0.kinesis_stream.#: "0",
outputs.0.lambda.#: "1",
outputs.0.lambda.0.resource_arn: "arn:aws:lambda:us-west-2:XXXXXXXXXXXX:function:some-lambda",
outputs.0.lambda.0.role_arn: "arn:aws:iam::XXXXXXXXXXXX:role/some-lambda",
outputs.0.name: "DESTINATION_SQL_STREAM",
outputs.0.schema.#: "1",
outputs.0.schema.0.record_format_type: "JSON",
```

I'm wondering if there's something wrong in my configuration or it's not possible to create the connection to a stream created by Kinesis Analytics by default. Thanks!

**Comment by kl4w** [comment]
@netrebel there's nothing wrong with your configuration. For some reason, I didn't iterate over the list of outputs and only captures the first output. Can you please submit a new Issue if not yet created?

**Comment by netrebel** [comment]
@kl4w Before creating the issue I updated to the latest terraform (0.11.11) and aws provider (1.58) and when I ran it again and I got this in the log, which I had not seen before:

```
aws_kinesis_analytics_application.detection: Modifying... (ID: arn:aws:kinesisanalytics:us-west-2:XXXX...XXX:application/tf-vee2-vee2-detection)
  outputs.#:                             "1" => "2"
  outputs.1.lambda.#:                    "0" => "1"
  outputs.1.lambda.0.resource_arn:       "" => "arn:aws:lambda:us-west-2:XXXXXXXXXXXX:function:some-lambda"
  outputs.1.lambda.0.role_arn:           "" => "arn:aws:iam::XXXXXXXXXXXX:role/some-lambda-execution"
  outputs.1.name:                        "" => "error_stream"
  outputs.1.schema.#:                    "0" => "1"
  outputs.1.schema.0.record_format_type: "" => "JSON"
```

So I thought it had been fixed already, but looking at the AWS console I still don't see the connection to the `error_stream`. 

I think I see what you mean with taking only the first element:
https://github.com/kl4w/terraform-provider-aws/blob/ff41de43022c1f402f636829dddf9f5ce9d951ef/aws/resource_aws_kinesis_analytics_application.go#L1390-L1391

```golang
func flattenKinesisAnalyticsOutputs(outputs []*kinesisanalytics.OutputDescription) []interface{} {
	s := []interface{}{}

	if len(outputs) > 0 {
		id := outputs[0]
...
```
I'll create the Issue. Thanks for the quick response!
